src/concurrency/genmc/global_allocations.rs

```diff
@@ -0,0 +1,145 @@
+use std::collections::hash_map::Entry;
+use std::sync::RwLock;
+
+use genmc_sys::{GENMC_GLOBAL_ADDRESSES_MASK, getGlobalAllocStaticMask};
+use rand::rngs::StdRng;
+use rand::{Rng, SeedableRng};
+use rustc_const_eval::interpret::{
+    AllocId, AllocInfo, AllocKind, InterpResult, PointerArithmetic, interp_ok,
+};
+use rustc_data_structures::fx::FxHashMap;
+use rustc_middle::{err_exhaust, throw_exhaust};
+use tracing::info;
+
+use crate::alloc_addresses::align_addr;
+
+#[derive(Debug, Default)]
+pub struct GlobalAllocationHandler {
+    inner: RwLock<GlobalStateInner>,
+}
+
+/// This contains more or less a subset of the functionality of `struct GlobalStateInner` in `alloc_addresses`.
+#[derive(Clone, Debug)]
+struct GlobalStateInner {
+    // FIXME(genmc): Decide on an API for GenMC to access this (to check if access to global memory is valid), or decide we don't need that.
+    #[allow(unused)]
+    /// This is used as a map between the address of each allocation and its `AllocId`. It is always
+    /// sorted by address. We cannot use a `HashMap` since we can be given an address that is offset
+    /// from the base address, and we need to find the `AllocId` it belongs to. This is not the
+    /// *full* inverse of `base_addr`; dead allocations have been removed.
+    int_to_ptr_map: Vec<(u64, AllocId)>,
+    /// The base address for each allocation.
+    /// This is the inverse of `int_to_ptr_map`.
+    base_addr: FxHashMap<AllocId, u64>,
+    /// This is used as a memory address when a new pointer is casted to an integer. It
+    /// is always larger than any address that was previously made part of a block.
+    next_base_addr: u64,
+    /// To add some randomness to the allocations
+    /// FIXME(genmc): maybe seed this from the rng in MiriMachine?
```

Comment:
This is global state, not per-machine, so how's that supposed to work?
We could seed it with the `miri-seed`.